### PR TITLE
Unify pool config - part 1

### PR DESCRIPTION
**Pull request by michalwski**
This PR makes it possible to start following pools:
* redis
* riak
* cassandra
* http
* generic (used in sns, and push modules)

via `outgoing_pools` config option as described in #1959. The issue is not full covered by this PR yet. Remaining work relates to rdbms and elasticsearch pools.

Currently `big_tests` starts only riak and cassandra pools via the `outgoing_pools` option. Other pools while possible to start in this way are still started directly by modules using them.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[5581.1](https://travis-ci.org/esl/MongooseIM/jobs/428768166) / Erlang 19.3 / small_tests / 0c56fcf28bcfdc5cb9dc1ee4e3bf4faf8625e54d
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2077/5581/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2077/5581/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-14_19.41.39/lib.mongooseim.logs/run.2018-09-14_19.41.40/suite.log.html)

**Comment by mongoose-im** [comment]
[5590.1](https://travis-ci.org/esl/MongooseIM/jobs/429545339) / Erlang 19.3 / small_tests / a1132a537f7f7d7acfc64870db7a78bd31069384
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2077/5590/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2077/5590/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-17_14.11.48/lib.mongooseim.logs/run.2018-09-17_14.11.49/suite.log.html)

File: src/mod_push_service_mongoosepush.erl [hunk 1]
```diff
@@ -51,8 +51,9 @@ start(Host, Opts) ->
 
     MaxHTTPConnections = gen_mod:get_opt(max_http_connections, Opts, 100),
     mongoose_wpool:ensure_started(),
-    {ok, _} = mongoose_wpool:start(?MODULE, Host, [{strategy, available_worker},
-                                                   {workers, MaxHTTPConnections}]),
+    {ok, _} = mongoose_wpool:start(generic, Host, mod_push,
```
Comment (kzemek) [inline src/mod_push_service_mongoosepush.erl]: I kind of dislike replacing `?MODULE` with tags that are meant to reference the module but in a subtler way. If we see a pool process with e.g. `pusher_push` in its name, we'll grep around the code to find where it comes from. If the tag was simply `?MODULE`, everything would be clear from the getgo. In this instance, `mod_push` does not really correspond to a pool for `mongoosepush` service.

File: src/mongoose_riak.erl [hunk 2]
```diff
@@ -55,26 +55,20 @@ start() ->
...
-    {_, RiakAddr} = get_riak_opt(address, RiakOpts),
-    {_, RiakPort} = get_riak_opt(port, RiakOpts),
-    {_, Workers} = get_riak_opt(pool_size, RiakOpts, {pool_size, 20}),
-    SecurityOptsKeys = [credentials, cacertfile, ssl_opts],
-    SecurityOpts = [get_riak_opt(OptKey, RiakOpts) || OptKey <- SecurityOptsKeys],
-    RiakPBOpts = [auto_reconnect, keepalive],
-    WorkerArgs = maybe_add_additional_opts(RiakPBOpts, SecurityOpts),
-    Worker = {riakc_pb_socket, [RiakAddr, RiakPort, WorkerArgs]},
+    {_, Workers} = mongoose_wpool_riak:get_riak_opt(pool_size, RiakOpts, {pool_size, 20}),
     %% TODO (Bartek Gorny): improve worker_pool, then use available_worker strategy here
```
Comment (kzemek) [inline src/mongoose_riak.erl]: @bartekgorny I feel this comment has not aged well. What does it mean? Why would we want to use `available_worker`?

File: src/mongoose_wpool.erl [hunk 3]
```diff
@@ -35,16 +53,39 @@ ensure_started() ->
...
 
 start(Type, Host, Tag, PoolOpts) ->
-    {Opts0, WpoolOpts} = proplists:split(PoolOpts, [strategy, call_timeout]),
+    start(Type, Host, Tag, PoolOpts, []).
+
+start(Type, Host, Tag, PoolOpts, ConnOpts) ->
+    {Opts0, WpoolOptsIn} = proplists:split(PoolOpts, [strategy, call_timeout]),
     Opts = lists:append(Opts0),
-    case wpool:start_sup_pool(make_pool_name(Type, Host, Tag), WpoolOpts) of
+    %WpoolOpts = maybe_get_wpool_opts_from_callback(Type, WpoolOptsIn, ConnOpts),
```
Comment (kzemek) [inline src/mongoose_wpool.erl]: Leftovers?

File: src/mongoose_wpool.erl [hunk 4]
```diff
@@ -137,8 +184,60 @@ get_pools() ->
...
+          ?ERROR_MSG("event=wpool_callback_init_error, error=~p, reason=~p, stacktrace=~p",
+                     [E, R, erlang:get_stacktrace()]),
+          {error, pool_not_initialized}
+    end.
+
+stop_by_callback(Type, Host, Tag) ->
+    try
+        CallbackModule = make_callback_module_name(Type),
+        CallbackModule:stop(Host, Tag)
+    catch E:R ->
```
Comment (kzemek) [inline src/mongoose_wpool.erl]: Feels like this try/catch dance could be deduplicated

File: src/mod_push_service_mongoosepush.erl (hunk 1 above)
Comment (michalwski) [reply src/mod_push_service_mongoosepush.erl]: 1. Previously `?MODULE` was used as the pool's type which doesn't fit nicely to the convention introduced and explained in #1959.
2. On one hand using ?MODULE as a tag name would be handy, on the other hand in the future the tag name is supposed to be configurable, which makes greping even more difficult.
3. I'm ok with using `?MODULE` as a tag for pool which are not started via `outgoing_pools` option.

@fenek I think your input on this is needed.

File: doc/advanced-configuration/outgoing-connections.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    * `global` - meaning the pool will started once despited the XMPP hosts served by MongooseIM
+    * `host` - the pool will be started for all the XMPP hosts served by MongooseIM
+    * a binary representing a specific XMPP host like `<<"domain1.chat.im">>`
+* `Tag` is a name to distinguish pools with the same `Type` and `Host` parameter.
+* `PoolOptions` is a list of `{key, value}` pairs as defined in [worker_pool doc](https://github.com/inaka/worker_pool#choosing-a-strategy)
+   with the following exception:
+    * `strategy` - specifies worker selection strategy for the given pool, default is `best_worker`,
+      more details on this can be found in [Choosing strategy in worker_pool doc](https://github.com/inaka/worker_pool#choosing-a-strategy)
+    * `call_timeout` - specifies the timeout for a call operation to the pool
+* `ConnectionOptions` - list of `{key, value}` pairs passed to the `start` function of the pool type
```
[the remark this answers is not in the document]
Comment (michalwski) [reply doc/advanced-configuration/outgoing-connections.md]: Good catch!

File: src/mongoose_wpool.erl (hunk 3 above)
Comment (michalwski) [reply src/mongoose_wpool.erl]: Yes, leftovers :)

File: src/mongoose_wpool.erl (hunk 4 above)
Comment (michalwski) [reply src/mongoose_wpool.erl]: I'll try to deduplicate it.

**Comment by mongoose-im** [comment]
[5603.1](https://travis-ci.org/esl/MongooseIM/jobs/430009740) / Erlang 19.3 / small_tests / 55bc8757a8975449b358535743325c516fac5f38
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2077/5603/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2077/5603/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-18_11.57.06/lib.mongooseim.logs/run.2018-09-18_11.57.07/suite.log.html)

**Comment by mongoose-im** [comment]
[5605.1](https://travis-ci.org/esl/MongooseIM/jobs/430032533) / Erlang 19.3 / small_tests / ac51acec933aa101c226018cb065a60fd5ac7d48
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2077/5605/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2077/5605/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-18_13.24.23/lib.mongooseim.logs/run.2018-09-18_13.24.24/suite.log.html)

File: src/mod_push_service_mongoosepush.erl (hunk 1 above)
Comment (kzemek) [reply src/mod_push_service_mongoosepush.erl]: I'd be fine with merging this as-is if we only change the `mod_push` tag to something a little more fitting.

**Comment by mongoose-im** [comment]
[5610.1](https://travis-ci.org/esl/MongooseIM/jobs/430410740) / Erlang 19.3 / small_tests / 8936dc5d1c3b910fb6c63bf26592164394b077d4
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2077/5610/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2077/5610/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-19_07.29.24/lib.mongooseim.logs/run.2018-09-19_07.29.25/suite.log.html)

File: doc/advanced-configuration/outgoing-connections.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+```erlang
+{Type, Host, Tag, PoolOptions, ConnectionOptions}
+```
+
+Where:
+
+* `Type` is one of the types listed above
+* `Host` can be set to:
+    * `global` - meaning the pool will started once despited the XMPP hosts served by MongooseIM
```
Comment (goddammit) [inline doc/advanced-configuration/outgoing-connections.md]: `global` - meaning the pool will be started once no matter how many XMPP hosts are served by MongooseIM

**Pull request merged**